common/djangoapps/entitlements/utils.py

```diff
@@ -58,3 +59,36 @@ def is_course_run_entitlement_fulfillable(
     can_upgrade = unexpired_paid_modes and entitlement.mode in unexpired_paid_modes
 
     return course_overview.start and can_upgrade and (is_enrolled or can_enroll)
+
+
+def revoke_entitlements_and_downgrade_courses_to_audit(course_entitlements, user_id, awarded_cert_course_ids):
+    """
+    This method expires the entitlements for provided course_entitlements and also moves the enrollments
+    to audit for the course entitlements which have not been completed yet(not a part of the provided exclusion_list).
+    """
+    log.info("B2C_SUBSCRIPTIONS: Starting revoke_entitlements_and_downgrade_courses_to_audit for "
+             "user_id: %s and course_entitlements %s",
+             user_id,
+             course_entitlements)
+    for course_entitlement in course_entitlements:
+        if course_entitlement.enrollment_course_run is None:
+            if course_entitlement.expired_at is None:
+                course_entitlement.expire_entitlement()
+
+        elif course_entitlement.enrollment_course_run.course_id not in awarded_cert_course_ids:
+            course_id = course_entitlement.enrollment_course_run.course_id
+            enrollment_mode = course_entitlement.enrollment_course_run.mode
+            username = course_entitlement.enrollment_course_run.user.username
+            if enrollment_mode == CourseMode.VERIFIED:
+                course_entitlement.set_enrollment(None)
+                if course_entitlement.expired_at is None:
+                    course_entitlement.expire_entitlement()
+                update_enrollment(username, str(course_id), CourseMode.AUDIT, include_expired=True)
+            else:
+                log.warning("B2C_SUBSCRIPTIONS: Enrollment mode mismatch for user_id: %s and course_id: %s",
+                            user_id,
+                            course_id)
+    log.info("B2C_SUBSCRIPTIONS: Completed revoke_entitlements_and_downgrade_courses_to_audit for "
+             "user_id: %s and course_entitlements %s",
+             user_id,
+             course_entitlements)
```

Comment:
`course_entitlements` here is an ORM instance. If you want to log the list of users' revokable entitlements, please use a flat list or remove this altogether.